### PR TITLE
[L1T] Phase-2, correlator developments

### PR DESCRIPTION
**Pull request by cecilecaillol**
#### PR description:

Developments to phase-2 L1 correlator. Porting of local https://github.com/cms-l1t-offline/cmssw/pull/1031.

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

File: L1Trigger/Phase2L1ParticleFlow/interface/egamma/pftkegsorter_ref.h [hunk 1]
```diff
@@ -5,12 +5,10 @@
 #include <vector>
 
 #include "DataFormats/L1TParticleFlow/interface/layer1_emulator.h"
+#include "L1Trigger/Phase2L1ParticleFlow/interface/dbgPrintf.h"
 
 #ifdef CMSSW_GIT_HASH
```
Comment (AdrianoDee) [inline L1Trigger/Phase2L1ParticleFlow/interface/egamma/pftkegsorter_ref.h:10]: For my understanding, why is this needed?

File: L1Trigger/Phase2L1ParticleFlow/test/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+*.txt
```
Comment (AdrianoDee) [inline L1Trigger/Phase2L1ParticleFlow/test/.gitignore:1]: I'm not sure about the policy here but I would remove this file.

File: L1Trigger/Phase2L1ParticleFlow/src/regionizer/multififo_regionizer_ref.cpp [hunk 3]
```diff
@@ -72,15 +72,13 @@ l1ct::MultififoRegionizerEmulator::MultififoRegionizerEmulator(unsigned int nend
...
-          caloRoutes_.emplace_back(is + 3 * ie, il, 3 * is + j + 9 * ie, il);
-          if (j == 0 || j == 2) {
-            int other = (j == 0) ? 2 : 1;  // pf region 0, takes from prev. pf region 2 takes from next
-            //                       from sector      , from link, to region, to fifo
-            caloRoutes_.emplace_back(
-                (is + other) % 3 + 3 * ie, il, 3 * is + j + 9 * ie, il + 2);  //last 2 = NCALOFIBERS
+    for (unsigned int is = 0; is < NCALO_SECTORS; ++is) {                      // NCALO_SECTORS sectors
+      for (unsigned int il = 0; il < NCALO_LINKS; ++il) {                      // max clusters per sector per clock
+        for (unsigned int j = 0; j < 3; ++j) {                                 // PF REGION
+          caloRoutes_.emplace_back(is + 3 * ie, il, 3 * is + j + 9 * ie, il);  // 4 args are: sector, link, region, fifo
```
Comment (AdrianoDee) [inline L1Trigger/Phase2L1ParticleFlow/src/regionizer/multififo_regionizer_ref.cpp:78]: Seems these numbers (and the ones below at line 81) are `NCALO_SECTORS` and `NTK_SECTORS`, aren't them? I would use the names or define them somewhere if I misunderstood just to make this more readable.

File: L1Trigger/Phase2L1ParticleFlow/src/L1TCorrelatorLayer1PatternFileWriter.cc [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+#include <iostream>
+
+L1TCorrelatorLayer1PatternFileWriter::L1TCorrelatorLayer1PatternFileWriter(const edm::ParameterSet& iConfig,
+                                                                           const l1ct::Event& eventTemplate)
+    : partition_(parsePartition(iConfig.getParameter<std::string>("partition"))),
+      writeInputs_(iConfig.existsAs<std::string>("inputFileName") &&
+                   !iConfig.getParameter<std::string>("inputFileName").empty()),
+      writeOutputs_(iConfig.existsAs<std::string>("outputFileName") &&
+                    !iConfig.getParameter<std::string>("outputFileName").empty()),
+      outputBoard_(-1),
```
Comment (AdrianoDee) [inline L1Trigger/Phase2L1ParticleFlow/src/L1TCorrelatorLayer1PatternFileWriter.cc:12]: Are these "constant parameters" needed in future? Because if not I'm not understanding why to define them as parameters and not as constants (e.g. in the class definition).
